File: src/agents/audit_chain.py
"""
AuditChain — Immutable trace of thought structures and decisions.

Anchoring: SHA256 + audit_chain: active
Validation: self-aware echo, causal consistency
"""

import hashlib
import json
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any


@dataclass
class AuditEntry:
    """Single immutable entry in the audit chain."""

    timestamp: str
    intent: str
    pattern: str
    decision: str
    outcome: str | None
    prev_hash: str
    entry_hash: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "timestamp": self.timestamp,
            "intent": self.intent,
            "pattern": self.pattern,
            "decision": self.decision,
            "outcome": self.outcome,
            "prev_hash": self.prev_hash,
            "entry_hash": self.entry_hash,
        }


class AuditChain:
    """
    Immutable audit chain for traceable thought structures.
    Each entry links to the previous via hash — causal consistency.
    """

    def __init__(self, anchor: str | None = None):
        self.anchor = anchor or "audit_chain: active"
        self._chain: list[AuditEntry] = []
        self._last_hash = ""
        self._outcomes: dict[str, str] = {}  # entry_hash -> outcome (reentrant learning)

    def _compute_hash(self, data: dict[str, Any], prev_hash: str) -> str:
        payload = json.dumps(data, sort_keys=True) + prev_hash
        return hashlib.sha256(payload.encode()).hexdigest()

    def append(
        self,
        intent: str,
        pattern: str,
        decision: str,
        outcome: str | None = None,
    ) -> AuditEntry:
        """Append a new audit entry. Returns the created entry."""
        timestamp = datetime.utcnow().isoformat() + "Z"
        data = {
            "timestamp": timestamp,
            "intent": intent,
            "pattern": pattern,
            "decision": decision,
            "outcome": outcome,
        }
        entry_hash = self._compute_hash(data, self._last_hash)

        entry = AuditEntry(
            timestamp=timestamp,
            intent=intent,
            pattern=pattern,
            decision=decision,
            outcome=outcome,
            prev_hash=self._last_hash,
            entry_hash=entry_hash,
        )
        self._chain.append(entry)
        self._last_hash = entry_hash
        return entry

    def verify(self) -> bool:
        """Verify causal consistency of the chain."""
        prev_hash = ""
        for entry in self._chain:
            data = {
                "timestamp": entry.timestamp,
                "intent": entry.intent,
                "pattern": entry.pattern,
                "decision": entry.decision,
                "outcome": entry.outcome,
            }
            expected = self._compute_hash(data, prev_hash)
            if expected != entry.entry_hash:
                return False
            prev_hash = entry.entry_hash
        return True

    def attach_outcome(self, entry_hash: str, outcome: str) -> None:
        """Attach outcome for reentrant learning. Preserves chain immutability."""
        self._outcomes[entry_hash] = outcome

    def export_trace(self) -> list[dict[str, Any]]:
        """Export full trace for external audit (includes learned outcomes)."""
        result = []
        for e in self._chain:
            d = e.to_dict()
            if e.entry_hash in self._outcomes:
                d["learned_outcome"] = self._outcomes[e.entry_hash]
            result.append(d)
        return result

    def __len__(self) -> int:
        return len(self._chain)

File: genesis-sovereign-fabric/crates/gsf-core/src/audit_chain.rs
//! Immutable AuditChain. SHA256-verkettet.
//! Genesis: sha256:acb92fd8346a65ff17dbf9a41e3003f2d566a17f839af4c3a90a4b4b1789dd28a

use sha2::{Digest, Sha256};
use serde::{Deserialize, Serialize};

pub const GENESIS_ANCHOR: &str = "acb92fd8346a65ff17dbf9a41e3003f2d566a17f839af4c3a90a4b4b1789dd28a";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuditEntry {
    pub timestamp: String,
    pub intent: String,
    pub pattern: String,
    pub decision: String,
    pub outcome: Option<String>,
    pub prev_hash: String,
    pub entry_hash: String,
}

pub struct AuditChain {
    chain: Vec<AuditEntry>,
    last_hash: String,
}

impl AuditChain {
    pub fn new() -> Self {
        Self {
            chain: Vec::new(),
            last_hash: GENESIS_ANCHOR.to_string(),
        }
    }

    pub fn append(
        &mut self,
        intent: &str,
        pattern: &str,
        decision: &str,
        outcome: Option<&str>,
    ) -> AuditEntry {
        let timestamp = chrono::Utc::now().to_rfc3339();
        let data = format!(
            "{}|{}|{}|{}|{}|{}",
            timestamp,
            intent,
            pattern,
            decision,
            outcome.unwrap_or(""),
            self.last_hash
        );
        let entry_hash = hex::encode(Sha256::digest(data.as_bytes()));
        let entry = AuditEntry {
            timestamp,
            intent: intent.to_string(),
            pattern: pattern.to_string(),
            decision: decision.to_string(),
            outcome: outcome.map(String::from),
            prev_hash: self.last_hash.clone(),
            entry_hash: entry_hash.clone(),
        };
        self.last_hash = entry_hash;
        self.chain.push(entry.clone());
        entry
    }

    pub fn verify(&self) -> bool {
        let mut prev = GENESIS_ANCHOR;
        for e in &self.chain {
            let data = format!(
                "{}|{}|{}|{}|{}|{}",
                e.timestamp,
                e.intent,
                e.pattern,
                e.decision,
                e.outcome.as_deref().unwrap_or(""),
                prev
            );
            let expected = hex::encode(Sha256::digest(data.as_bytes()));
            if expected != e.entry_hash {
                return false;
            }
            prev = &e.entry_hash;
        }
        true
    }

    pub fn export(&self) -> &[AuditEntry] {
        &self.chain
    }

    pub fn restore_entry(&mut self, entry: AuditEntry) {
        self.last_hash = entry.entry_hash.clone();
        self.chain.push(entry);
    }

    pub fn set_last_hash(&mut self, h: &str) {
        self.last_hash = h.to_string();
    }
}
